waspc/cli/src/Wasp/Cli/Command/Build.hs

```diff
@@ -58,40 +73,65 @@ build = do
-    --     users  (no changes in CI/CD scripts).
-    -- For more details, read the issue linked above.
-    copyUserFilesNecessaryForBuild waspProjectDir buildDir = do
-      copyDirectory
-        (waspProjectDir </> srcDirInWaspProjectDir)
-        (buildDir </> castRel srcDirInWaspProjectDir)
+    prepareFilesNecessaryForDockerBuild waspProjectDir buildDir = runExceptT $ do
+      -- Until we implement the solution described in https://github.com/wasp-lang/wasp/issues/1769,
+      -- we're copying all files and folders necessary for the build into the .wasp/build directory.
+      -- We chose this approach for 0.12.0 (instead of building from the project root) because:
+      --   - The build context remains small (~1.5 MB vs ~900 MB).
+      --   - We don't risk copying possible secrets from the project root into the build context.
+      --   - The commands for building the project stay the same as before
+      --     0.12.0, which is good for both us (e.g., for fly deployment) and our
+      --     users  (no changes in CI/CD scripts).
+      -- For more details, read the issue linked above.
+      liftIO $
+        copyDirectory
+          (waspProjectDir </> srcDirInWaspProjectDir)
+          (buildDir </> castRel srcDirInWaspProjectDir)
+
+      liftIO $
+        copyDirectory
+          (waspProjectDir </> dotWaspDirInWaspProjectDir </> generatedCodeDirInDotWaspDir </> sdkRootDirInGeneratedCodeDir)
+          (buildDir </> sdkRootDirInGeneratedCodeDir)
+
+      let packageJsonInBuildDir = buildDir </> castRel packageJsonInWaspProjectDir
+      let packageLockJsonInBuildDir = buildDir </> castRel packageLockJsonInWaspProjectDir
+
+      liftIO $
+        copyFile
+          (waspProjectDir </> packageJsonInWaspProjectDir)
+          packageJsonInBuildDir
+
+      liftIO $
+        copyFile
+          (waspProjectDir </> packageLockJsonInWaspProjectDir)
+          packageLockJsonInBuildDir
 
-      copyDirectory
-        (waspProjectDir </> dotWaspDirInWaspProjectDir </> generatedCodeDirInDotWaspDir </> sdkRootDirInGeneratedCodeDir)
-        (buildDir </> sdkRootDirInGeneratedCodeDir)
+      -- A hacky quick fix for https://github.com/wasp-lang/wasp/issues/2368
+      -- We should remove this code once we implement a proper solution.
+      ExceptT $ updateJsonFile removeWaspConfigDevDependency packageJsonInBuildDir
+      ExceptT $ updateJsonFile removeAllMentionsOfWaspConfig packageLockJsonInBuildDir
 
-      copyFile
-        (waspProjectDir </> packageJsonInWaspProjectDir)
-        (buildDir </> castRel packageJsonInWaspProjectDir)
+    removeAllMentionsOfWaspConfig :: Value -> Value
+    removeAllMentionsOfWaspConfig packageLockJsonObject =
+      packageLockJsonObject
+        & key "packages" . key "" %~ removeWaspConfigDevDependency
+        & key "packages" . _Object
+          %~ HM.filterWithKey
+            (\packageKey _ -> not ("wasp-config" `isInfixOf` packageKey))
```

Comment:
Could you maybe explain what is going on here, maybe in a comment above the since it's quite hard to grasp.

"Here we are removing this from X and then we are doing ..."

Reply:
Yeah hm. I guess this is on us since we don't know lenses well, but they are advanced Haskell concept I guess, at least for our team at the moment.

Reading this, I get taht in object, under "packages" key, we remove Wasp config dev dependency, althogh that `key ""` confuses me, what is empty key hm? If there wasn't `key ""` in that first part, I think it would need no comment.

Second part also starts good, `key "packages"`, I can disregard `_Object` or take it as "ok, in the object", then we filter by keys I guess, ok then `packageKey` confuses me, I would either go with `key` or `packageName`, and then we say key has to have `wasp-config` as infix. Ok, so we filter/keep those without wasp-config as infix. This was actually readable, although I would change `packageKey` as I suggested + maybe just go with
```hs
HM.filterWithKey (not . ("wasp-config" `isInfixOf`))
```
to reduce the noise.

I would also add short comments though, just to be safe, but would also make code a bit more readable as suggested.

Btw I am always wary when I see usage of `isInfixOf` like this, or prefix or whatever, as often it is too wide of a net. Why `isInfixOf`? Why do you can't be more specific, I guess there are some extra characters in this string? But `isInfixOf` is bad because what if they package called `owasp-configuration` that has nothing to do with wasp? Yeah, doesn't seem great, I would make it more specific.

Reply:
> "Here we are removing this from X and then we are doing ..."

Sure, will do!

> althogh that key "" confuses me, what is empty key hm?

[Answer](https://youtu.be/WyYC9qRlJxY?t=205)

I'll add a comment for that too.

![image](https://github.com/user-attachments/assets/86d20f7a-2ebb-4786-af48-7157c4041bed)

> Btw I am always wary when I see usage of isInfixOf

What would you suggest? Here are the keys it has to remove:
```
../../data/packages/wasp-config
node_modules/wasp-config
/home/filip/.../wasp-config
<Other possible paths on user machines>
```
`isSuffixOf` with `/wasp-config` (notice the slash) might be enough, but arguably it's better to delete something extra than miss something, so keep that in mind (I don't know what might happen with a trailing slash, although it's safe to guess we won't have to deal with it).

Finally, keep in mind this is code we want to get rid of in a couple of weeks tops, so getting to clever here probably isn't worth it.

Either way, I'll document the strings I expect to match, you would have been able to give a concrete suggestion immediately if I'd done so in the first place.

> To reduce the noise:
> ```haskell
> HM.filterWithKey (not . ("wasp-config" `isInfixOf`))
>```

Hm, I don't think this will work.

I remember trying all kinds of compositions and functions, and lambda ended up the cleanest one that compiles. I think this one won't work because it doesn't take into account that the callback receives two arguments (key and value).

I managed to avoid the lambda with some combination of composition, `const`, and `uncurry`, but it was much less readable IMO.

>  I would either go with `key` or `packageName`

`packageName` would be incorrect (see examples of keys I listed above). `key` is ok, but sounded less informative than `packageKey`. I can change it to `key` if that's what you prefer, I'm fine either way.

> I would also add short comments though, just to be safe, but would also make code a bit more readable as suggested.

I'm sorry I can't resist... You know you're working in a state-of-the-art language when you need comments to understand code that changes a JSON property :smile: 

And yes, I agree we need comments, I'll add some.

Reply:
Oh wow it is actually a "" key, ok I didn't expect that, thought it is some kind of quirky thing you had to do to get the lenses to do what you want, but it is actually exactly what it looks like.

infixOf -> seems like you want to target a path that ends up with `/wasp-config`. So then go with that -> check if `/wasp-config` is a suffix. Not sure about the trailing slash? Add logic for that, can be just simple OR. Be careful not to hardcode file seprator, use the one from FilePath.</> so it is system agnostic (unless they are guaranteed to be posix by package.json, in that case go with posix).
This is temporary fix, but you how temporary fixes are, they can stay in longer than one would expect. Also, this is so simple and easy to fix, and final solution is equally simple, we are certainly not overengineering here. And, there is value in me pointing this out to you, because you are less likely to make the same mistake next time.
Better to delete too much than too little -> probably depends on the context, but I don't agree in this case, because this can result in some nasty bugs, in the sense that they are hard to figure out because this is super unexpected behaviour. And I don't think going with suffix we have much chance to miss something, I would call it strictly and obviously better solution with the same level of complexity.

That lambda with packageKey -> my bad, I missed that `_` there, thought it is just one argument, I agree not worth bothering with it.

`packageKey` -> ok, so those are paths actually from what I get? I mean what is that, how would you describe if I you had to? `key` is correct because it is key in the object. `packageKey` doesn't sound correct because I doubt that thing is called `package key`. Obviously also not a name, as you pointed out -> when I was suggesting that I didn't really know what is it. Now I would be tempted to call it `packagePath` or `packageUri` or something like that. I would go with `key` if you wish to keep it simple and correct, or I would check the npm docs, figure out what can be in that string + waht they call it, and call it like that.

I know you were making a joke regarding JSON and comments, but what lenses do here is quite impressive to me actually, you can't do taht in Javascript in one line of code while not mutating the origianl object, and spread operators can also be a pain to write when you have to go couple of levels deep, while ont being reusable or anything.